RecoLocalCalo/EcalRecAlgos/src/EcalUncalibRecHitTimingCCAlgo.cc

```diff
@@ -0,0 +1,167 @@
+      pulse.at(isample) = fullpulse(isample + offset);
+      pedSubSamples.at(isample) =
+          std::max(0., pedSubSamples.at(isample) - amplitudes[ipulse] * pulse.at(isample) / pulsenorm);
+    }
+  }
+
+  float tStart = startTime_ + GLOBAL_TIME_SHIFT;
+  float tStop = stopTime_ + GLOBAL_TIME_SHIFT;
+  float tM = (tStart + tStop) / 2;
+
+  float distStart, distStop;
+  int counter = 0;
+
+  do {
+    ++counter;
+    distStart = computeCC(pedSubSamples, fullpulse, tStart);
+    distStop = computeCC(pedSubSamples, fullpulse, tStop);
+
+    if (distStart > distStop) {
+      tStop = tM;
+    } else {
+      tStart = tM;
+    }
+    tM = (tStart + tStop) / 2;
+
+  } while (tStop - tStart > targetTimePrecision_ && counter < MAX_NUM_OF_ITERATIONS);
+
+  tM -= GLOBAL_TIME_SHIFT;
+
+  if (counter < MIN_NUM_OF_ITERATIONS || counter > MAX_NUM_OF_ITERATIONS - 1) {
+    if (counter > MAX_NUM_OF_ITERATIONS / 2)
+      //Produce a log if minimization took too long
+      edm::LogInfo("EcalUncalibRecHitTimingCCAlgo::computeTimeCC")
+          << "Minimization Counter too high: " << counter << std::endl;
+    tM = TIME_WHEN_NOT_CONVERGING * ecalPh1::Samp_Period;
+    //Negative error means that there was a problem with the CC
+    errOnTime = -targetTimePrecision_ / ecalPh1::Samp_Period;
+  }
+
+  return -tM / ecalPh1::Samp_Period;
+}
+
+FullSampleVector EcalUncalibRecHitTimingCCAlgo::interpolatePulse(const FullSampleVector& fullpulse, const float t) {
+  // t is in ns
+  int shift = t / ecalPh1::Samp_Period;
+  if (t < 0)
+    shift -= 1;
+  float timeShift = t - ecalPh1::Samp_Period * shift;
+  float tt = timeShift / ecalPh1::Samp_Period;
+
+  FullSampleVector interpPulse;
+  // 2nd poly with avg
+  unsigned int numberOfSamples = fullpulse.size();
+  for (unsigned int i = 1; i < numberOfSamples - 2; ++i) {
+    float a = 0.25 * tt * (tt - 1) * fullpulse[i - 1] + (0.25 * (tt - 2) - 0.5 * (tt + 1)) * (tt - 1) * fullpulse[i] +
+              (0.25 * (tt + 1) - 0.5 * (tt - 2)) * tt * fullpulse[i + 1] + 0.25 * (tt - 1) * tt * fullpulse[i + 2];
+    if (a > 0)
+      interpPulse[i] = a;
+    else
+      interpPulse[i] = 0;
+  }
+  interpPulse[0] = (0.25 * (tt - 2) - 0.5 * (tt + 1)) * ((tt - 1) * fullpulse[0]) +
+                   (0.25 * (tt + 1) + 0.5 * (tt - 2)) * tt * fullpulse[1] + 0.25 * tt * (tt - 1) * fullpulse[2];
+  interpPulse[numberOfSamples - 2] = 0.25 * tt * (tt - 1) * fullpulse[numberOfSamples - 3] +
+                                     (0.25 * (tt - 2) - 0.5 * (tt + 1)) * (tt - 1) * fullpulse[numberOfSamples - 2] +
+                                     (0.25 * (tt + 1) - 0.5 * (tt - 2)) * tt * fullpulse[numberOfSamples - 1];
+  interpPulse[numberOfSamples - 1] = 0.5 * tt * (tt - 1) * fullpulse[numberOfSamples - 2] -
+                                     (tt * tt - 1) * fullpulse[numberOfSamples - 1] +
+                                     (0.25 * (tt + 1) - 0.5 * (tt - 2)) * tt * fullpulse[numberOfSamples - 1];
```

Comment:
```suggestion
  interpPulse[0] = fac * fullpulse[0] +
                   (0.25 * (tt + 1) + 0.5 * (tt - 2)) * tt * fullpulse[1] + facM1orP2 * fullpulse[2];
  interpPulse[numberOfSamples - 2] = facM1orP2 * fullpulse[numberOfSamples - 3] +
                                     fac * fullpulse[numberOfSamples - 2] +
                                     facP1 * fullpulse[numberOfSamples - 1];
  interpPulse[numberOfSamples - 1] = 2* facM1orP2 * fullpulse[numberOfSamples - 2] -
                                     4* facM1orP2 * fullpulse[numberOfSamples - 1] +
                                     facP1 * fullpulse[numberOfSamples - 1];
```
- some simplification based on the already computed constants; and I think that it's shorter and more readable
- is the sign in `[1]` for ` interpPulse[0]` correct? It's the only one different

Reply:
Well spotted! interpPulse[0] has a very small impact on the result (it's far from the central BX) but it is an incredible catch!